### PR TITLE
Added checkpoints description. Resolves : Improve checkpoints documentation #19708

### LINKED ISSUE #19708: Improve checkpoints documentation
UnboundedSource.CheckpointMark should mention that implementations it should be encodable (have an associated Coder). Also maybe it is a good idea to explain a bit more the checkpointing semantics on Beam.

Imported from Jira [BEAM-7691](https://issues.apache.org/jira/browse/BEAM-7691). Original Jira may contain additional context.
Reported by: iemejia.

### PR DESCRIPTION
**Pull request by DevPJ9**
Description :

**Please** add a meaningful description for your change here

Added a the description point `<p>Note that: The implementations of this interface should be encodable (have an associated Coder)` about UnboundedSource.CheckpointMark. fixes #19708

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Review by damccorm — APPROVED** [review]
LGTM, thanks!

**Comment by damccorm** [comment]
Rerunning some flaky test suites

**Comment by damccorm** [comment]
Run Spotless PreCommit

**Comment by damccorm** [comment]
Run Java_Kinesis_IO_Direct PreCommit

**Comment by damccorm** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by damccorm** [comment]
Run Spotless PreCommit

**Comment by damccorm** [comment]
Run Java_Kinesis_IO_Direct PreCommit

**Comment by damccorm** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by DevPJ9** [comment]
> Sorry, coming back to this - we had a CI outage that kept me from merging before and distracted my attention

No problem at all, I completely understand. Thank you for keeping me informed about the situation.

**Comment by DevPJ9** [comment]
Sorry, I pressed the `close with comment` button by mistake.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java [hunk 1]
```diff
@@ -100,6 +100,8 @@ public boolean requiresDeduping() {
    * org.apache.beam.sdk.io.UnboundedSource.UnboundedReader}.
    *
    * <p>For example, this could be offsets in a set of files being read.
+   *
+   * <p>Note that: The implementations of this interface should be encodable (have an associated Coder).
```
[the remark this answers is not in the document]
Comment (DevPJ9) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java:104]: Ok. should I commit again after executing `./gradlew :sdks:java:core:spotlessApply` ?

[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java:104]: Yep, you can either do that locally and commit to the same branch, or you can just click `commit suggestion` this time.

**Comment by damccorm** [comment]
Run Spotless PreCommit

**Review by damccorm — APPROVED** [review]
Thanks for making the changes!

**Pull request merged**